Mes 7705 cat a mod 2

gittins7 opened the pull request:
## Description
E2E Practice flow for cat A Mod 2
## Checklist

- [x] PR title includes the JIRA ticket number
- [x] Branch is rebased against the latest develop
- [x] Code has been tested manually
- [x] PR link added to JIRA ticket
- [ ] One review from each scrum team
- [ ] Squashed commit contains the JIRA ticket number

## Screenshots (optional)

andrewsetterfield committed:
WIP

andrewsetterfield committed:
WIP

andrewsetterfield committed:
wip

andrewsetterfield committed:
wip

andrewsetterfield committed:
wip

gittins7 committed:
Cat A Mod2 E2E practice flow

gittins7 committed:
removed un-needed files

matthew2564 committed:
Update behaviour map for Mod2 to only show safety and balance when there is a value

matthew2564 commented on src/app/pages/office/cat-a-mod2/office.cat-a-mod2.page.ts:
```diff
@@ -1,35 +1,287 @@
+      faultSummaryProvider,
+      faultCountProvider,
+    );
+    this.form = new FormGroup({});
+    this.weatherConditions = this.weatherConditionProvider.getWeatherConditions();
```
These are created in this base page so can be removed

matthew2564 commented on src/app/pages/office/cat-a-mod2/office.cat-a-mod2.page.ts:
```diff
@@ -1,35 +1,287 @@
+export class OfficeCatAMod2Page extends OfficeBasePageComponent {
+  pageState: OfficePageState;
+  pageSubscription: Subscription;
+  form: FormGroup;
```
This can be removed

gittins7 replied:
I thought we needed the pageState as it consists of a few fields bespoke to mod2 spread with the commonOfficePageState?

matthew2564 replied:
I left the comment on the ```form: FormGroup;``` line

gittins7 replied:
ah sorry thought it was for lines 97-99

matthew2564 replied:
This has not been resolved

matthew2564 commented on src/app/pages/office/cat-a-mod2/office.cat-a-mod2.page.ts:
```diff
@@ -1,35 +1,287 @@
+  static readonly maxFaultCount = 10;
+
+  activityCodeOptions: ActivityCodeModel[];
+  conductedLanguage: string;
```
This appears to be unused

gittins7 replied:
maxFaultCount is used in shouldDisplayDrivingFaultComments(), have gotten rid of the other 2

matthew2564 replied:
I have left the comment on the ```conductedLanguage: string;``` line

matthew2564 committed:
wrtc and tr specs for mod2

gittins7 committed:
test fixes

gittins7 committed:
added accompaniment-card to office.cat-home-test.page.html as it was missing

added updateValueAndValidity() to passCertificateNumberChanged() for pass-certificate-number in order to fix validation

matthew2564 commented on src/app/pages/office/cat-a-mod2/office.cat-a-mod2.page.ts:
```diff
@@ -1,35 +1,287 @@
+    this.store$.dispatch(ModeOfTransportChanged(modeOfTransport));
   }
 
+  weatherConditionsChanged(weatherConditions: WeatherConditions[]): void {
+    this.store$.dispatch(WeatherConditionsChanged(weatherConditions));
+  }
+
+  routeNumberChanged(routeNumber: number) {
+    this.store$.dispatch(RouteNumberChanged(routeNumber));
+  }
+
+  candidateDescriptionChanged(candidateDescription: string) {
+    this.store$.dispatch(CandidateDescriptionChanged(candidateDescription));
+  }
+
+  additionalInformationChanged(additionalInformation: string): void {
+    this.store$.dispatch(AdditionalInformationChanged(additionalInformation));
+  }
```
I would imagine this could all be removed as they will exist in the office base page

matthew2564 commented on src/app/pages/pass-finalisation/cat-a-mod2/__tests__/pass-finalisation.cat-a-mod2.page.spec.ts:
```diff
@@ -1,31 +1,139 @@
     component = fixture.componentInstance;
-    fixture.detectChanges();
+    component.subscription = new Subscription();
+    store$ = TestBed.get(Store);
```
this should be ```TestBed.inject```

matthew2564 commented on src/app/pages/pass-finalisation/cat-a-mod2/pass-finalisation.cat-a-mod2.page.ts:
```diff
@@ -1,29 +1,118 @@
-  ngOnInit() {
+    this.merged$ = merge(
+      transmission$.pipe(map((value) => this.transmission = value)),
+      testCategory$.pipe(map((value) => this.testCategory = value as TestCategory)),
```
Do we need ```testCategory$```? If not, we can also remove ```testCategory``` on line 48

gittins7 replied:
looks like it's passed into <bike-category-type> component in the html

matthew2564 replied:
```testCategory$``` exists in the base page, why would we not just do ```pageState.testCategory$ | async``` in the template?

matthew2564 commented on src/app/providers/category-whitelist/category-whitelist.ts:
```diff
@@ -14,7 +14,7 @@ export const whiteList: TestCategory[] = [
   TestCategory.CEM,
   TestCategory.C1EM,
-  // TestCategory.CCPC,
+  TestCategory.CCPC,
```
This can be undone

matthew2564 commented on src/store/tests/test-data/cat-a-mod2/safety-questions/safety-questions.cat-a-mod2.action.ts:
```diff
@@ -0,0 +1,12 @@
+import { createAction } from '@ngrx/store';
+import { QuestionOutcome } from '@dvsa/mes-test-schema/categories/common';
+
+export const SafetyQuestionOutcomeChanged = createAction(
+  '[VehicleChecksPage] [CatAMod2] Safety Question Outcome Changed',
+  (safetyQuestionOutcome: QuestionOutcome, index: number) => ({ safetyQuestionOutcome, index }),
+);
+
+export const AddSafetyQuestionComment = createAction(
+  '[Vehicle Checks] [CatAMod2] Add SafetyQuestion comment',
+  (comment: string) => ({ comment }),
+);
```
These (and more) already exist here ```src/store/tests/test-data/cat-a-mod2/safety-and-balance/safety-and-balance.cat-a-mod2.actions.ts```

matthew2564 commented on src/store/tests/test-data/cat-a-mod2/safety-questions/safety-questions.cat-a-mod2.selector.ts:
```diff
@@ -0,0 +1,26 @@
+import { CatDUniqueTypes } from '@dvsa/mes-test-schema/categories/D';
+import { CatD1UniqueTypes } from '@dvsa/mes-test-schema/categories/D1';
+import { CatDEUniqueTypes } from '@dvsa/mes-test-schema/categories/DE';
+import { CatD1EUniqueTypes } from '@dvsa/mes-test-schema/categories/D1E';
+import { SafetyQuestionResult } from '@dvsa/mes-test-schema/categories/common';
+import { createFeatureSelector } from '@ngrx/store';
+import { some } from 'lodash';
+
+export type CatAMod2SafetyQuestions =
+  | CatDUniqueTypes.SafetyQuestions
+  | CatD1UniqueTypes.SafetyQuestions
+  | CatDEUniqueTypes.SafetyQuestions
+  | CatD1EUniqueTypes.SafetyQuestions;
+
+export const getSafetyQuestions = (
+  safetyQuestionsCatDReducer: CatAMod2SafetyQuestions,
+): SafetyQuestionResult[] => {
+  return safetyQuestionsCatDReducer.questions;
+};
+
+export const safetyQuestionsExist = (safetyQuestions: CatAMod2SafetyQuestions): boolean => {
+  const questions = [...safetyQuestions.questions];
+  return some(questions, (fault) => fault.outcome != null);
+};
+
+export const getSafetyQuestionsCatD = createFeatureSelector<CatAMod2SafetyQuestions>('safetyQuestions');
```
This is all referencing Cat D, please remove all of this

matthew2564 commented on src/store/tests/test-summary/cat-a-mod2/test-summary.cat-a-mod2.actions.ts:
```diff
@@ -5,3 +5,5 @@ export const ModeOfTransportChanged = createAction(
   (modeOfTransport: ModeOfTransport) => ({ modeOfTransport }),
 );
+
+export const MODE_OF_TRANSPORT_CHANGED = '[Test Summary] [CatAMod2] Mode of transport changed';
```
Cant imagine why this would be needed, please remove

matthew2564 commented on src/store/tests/test-summary/cat-a-mod2/test-summary.cat-a-mod2.selector.ts:
```diff
@@ -1,3 +1,18 @@
+} from '@dvsa/mes-test-schema/categories/AM2';
 
 export const getModeOfTransport = (ts: TestSummary): ModeOfTransport => ts.modeOfTransport;
+export const getRouteNumber = (ts: TestSummary): number => ts.routeNumber;
+export const getCandidateDescription = (ts: TestSummary): string => ts.candidateDescription;
+export const getAdditionalInformation = (ts: TestSummary): string => ts.additionalInformation;
+export const getD255 = (ts: TestSummary): boolean => ts.D255;
+export const getIdentification = (ts: TestSummary): Identification => ts.identification;
+export const getSatNavUsed = (ts: TestSummary): boolean => ts.independentDriving === 'Sat nav';
+export const getTrafficSignsUsed = (ts: TestSummary): boolean => ts.independentDriving === 'Traffic signs';
+export const isDebriefWitnessed = (ts: TestSummary): boolean => ts.debriefWitnessed;
+export const getWeatherConditions = (ts: TestSummary): WeatherConditions[] => ts.weatherConditions;
+export const getIndependentDriving = (ts: TestSummary): IndependentDriving => ts.independentDriving;
```
These all exist here ```src/store/tests/test-summary/test-summary.selector.ts```

gittins7 committed:
merge comment amendments

gittins7 committed:
merge comment amendments

gittins7 committed:
reverted incorrect deletion

gittins7 committed:
test fixes

gittins7 committed:
removed testCategory$ as it already exists in the basepage for pass-finalisation

matthew2564 commented on src/app/pages/office/cat-a-mod2/components/safety-and-balance/__tests__/safety-and-balance.cat-a-mod2.spec.ts:
```diff
@@ -0,0 +1,131 @@
+      store$.dispatch(BalanceQuestionOutcomeChanged('P', index));
+    });
+
+    translate = TestBed.get(TranslateService);
```
Need to use .inject instead of .get

gittins7 committed:
removed testCategory$ as it already exists in the basepage for pass-finalisation

matthew2564 commented on src/app/shared/classes/test-flow-base-pages/office/office-base-page.ts:
```diff
@@ -506,12 +506,6 @@ export abstract class OfficeBasePageComponent extends PracticeableBasePageCompon
     ).subscribe();
   }
 
-  ionViewDidLeave(): void {
```
Why has this been removed?

gittins7 replied:
must have mistakenly removed it earlier on, will re-introduce

matthew2564 commented on src/app/shared/classes/test-flow-base-pages/office/__tests__/office-base-page.spec.ts:
```diff
@@ -280,15 +280,6 @@ fdescribe('OfficeBasePageComponent', () => {
     });
   });
 
-  describe('ionViewDidLeave', () => {
```
Why has this been removed?

gittins7 replied:
removed due to the absence of the method in the base page, will reintroduce with the actual method

matthew2564 commented on src/app/pages/office/cat-a-mod2/components/safety-and-balance/__tests__/safety-and-balance.cat-a-mod2.spec.ts:
```diff
@@ -0,0 +1,131 @@
+    });
+  });
+
+  beforeEach(async(() => {
```
async (in this context) has been deprecated in favour of waitForAsync

gittins7 committed:
merge comment amendments

gittins7 committed:
removed unused formGroup declaration
